### PR TITLE
Status page for a health-check of server dependencies. [#1342898]

### LINKED ISSUE #196: As an ops person I would like a health check URL
In order to do easy status and health checking for out instance, it would be convenient to have a url that returns some json and a relevant HTTP status code after checking that everything is working as expected (DB, Redis, ES, etc all available to the app).

### PR DESCRIPTION
**Pull request by ShawnMilo**
Currently covers:
- PostgreSQL
- Redis
- Elasticsearch

closes #196 

Sample output:

``` json
{
    "elasticsearch": {
        "response_ms": 44852,
        "status": "up",
        "status_code": 200
    },
    "postgresql": {
        "response_ms": 1580,
        "status": "up"
    },
    "redis": {
        "response_ms": 599,
        "status": "up",
        "uptime_in_seconds": 9,
        "used_memory": 878312,
        "used_memory_peak": 930208
    }
}
```

### CODE REVIEW

**Comment by ShawnMilo** [comment]
@cg: FYI

**Comment by pdpinch** [comment]
@bdero or @carsongee can you review?

File: .travis.yml [hunk 1]
```diff
@@ -3,10 +3,14 @@ python:
   - '2.7'
 services:
   - elasticsearch
+  - redis-server
 before_script:
   - sleep 10
 env:
-  - DATABASE_URL=postgres://postgres@localhost:5432/postgres
+  global:
```
Comment (carsongee) [inline .travis.yml:10]: Why did you add global? We aren't in a build matrix, so shouldn't need to specify that.

File: status/tests/test_status.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+from django.test.testcases import TestCase
+
+log = logging.getLogger(__name__)
+
+
+class TestStatus(TestCase):
+    """Test output of status page."""
+
+    def setUp(self):
+        """Set up client."""
```
Comment (carsongee) [inline status/tests/test_status.py]: Pleas add a docstring as to what you are doing here, and maybe a note saying, this saves and restores settings x,y,z in the class docstring so people know what is ok to mess with and what isn't later on.

File: status/views.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+log = logging.getLogger(__name__)
+
+
+def get_pg_info():
+    """Check PostgreSQL connection."""
+    try:
+        conf = settings.DATABASES['default']
+    except (AttributeError, KeyError):
+        log.error("No PostgreSQL connection info found in settings.")
+        return {"status": "no config found"}
```
Comment (carsongee) [inline status/views.py]: Please move all the messages to a shared dictionary or tuple to reduce chance of error and ease the ability to change status messages across services

File: status/views.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    try:
+        start = datetime.now()
```
Comment (carsongee) [inline status/views.py:49]: Please use https://docs.python.org/2/library/timeit.html for the time to increase resolution when possible

File: status/views.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+from redis import StrictRedis, ConnectionError as RedisConnectionError
+
+log = logging.getLogger(__name__)
+
+
+def get_pg_info():
+    """Check PostgreSQL connection."""
+    try:
+        conf = settings.DATABASES['default']
+    except (AttributeError, KeyError):
```
Comment (carsongee) [inline status/views.py:42]: I actually want these to raise up.  We shouldn't be catching anything in a health check in general, since the status code is what most health checkers use to determine health (the json is just a nice to have, but the status code is the critical thing)

File: status/views.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        log.error("No PostgreSQL connection info found in settings.")
+        return {"status": "no config found"}
+    try:
+        start = datetime.now()
+        connect(
+            database=conf["NAME"], user=conf["USER"], host=conf["HOST"],
+            port=conf["PORT"], password=conf["PASSWORD"]
+        )
+        micro = (datetime.now() - start).microseconds
+    except (OperationalError, KeyError):
```
Comment (carsongee) [inline status/views.py:57]: again, 500 preferred, so not catching

File: status/views.py (hunk 4 above)
Comment (carsongee) [inline status/views.py:49]: I think all these timers would be better done with a context manager, it will reduce repeated code and let us do a `finally` statement for these

File: status/views.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        return {"status": "down"}
+    return {"status": "up", "response_ms": micro}
+
+
+def get_redis_info():
+    """Check Redis connection."""
+    pat = re.compile(r'redis://([^:]+):(\d+)/(\d+)$')
+    try:
+        url = settings.BROKER_URL
+    except AttributeError:
```
Comment (carsongee) [inline status/views.py:69]: same on the exception advice, but I wonder if all this parsing is needed, surely there must be something in celery that already does this that we can reuse

File: status/views.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        url = settings.HAYSTACK_CONNECTIONS["default"]["URL"]
+    except (AttributeError, KeyError):
+        log.error("No Elasticsearch connection info in settings.")
+        return {"status": "no config found"}
+    match = pat.match(url)
+    if match is None:
+        log.error("Invalid Elasticsearch connection info: %s.", url)
+        return {"status": "no config found"}
+    host, port = match.groups()
+    search = Elasticsearch(hosts=[{"host": host, "port": port}])
```
Comment (carsongee) [inline status/views.py]: Elasticsearch can just take the string as it is in haystack so you don't need to parse it at all.  However, I would check the `HAYSTACK_CONNECTIONS` default and make sure it is of the ElasticSearch type above before testing, so that we could change it to solr for example, and this check wouldn't run

File: status/views.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    start = datetime.now()
+    try:
+        info = rdb.info()
+    except RedisConnectionError:
+        return {"status": "down"}
+    micro = (datetime.now() - start).microseconds
+    ret = {
+        "status": "up", "response_ms": micro,
+    }
+    fields = ("uptime_in_seconds", "used_memory", "used_memory_peak")
```
Comment (carsongee) [inline status/views.py:85]: Don't we need to close the connection?  We are connection limited by purchase and I would hate to have the connection pool go to infinity for health checks

File: status/views.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    try:
+        start = datetime.now()
+        connect(
+            database=conf["NAME"], user=conf["USER"], host=conf["HOST"],
+            port=conf["PORT"], password=conf["PASSWORD"]
+        )
+        micro = (datetime.now() - start).microseconds
+    except (OperationalError, KeyError):
+        log.error("Invalid PostgreSQL connection info in settings: %s", conf)
+        return {"status": "down"}
```
Comment (carsongee) [inline status/views.py]: Same here, I think we need a disconnect in a finally type statement to prevent connection pool exhaustion

File: status/views.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    micro = (datetime.now() - start).microseconds
+    return {
+        "status": "up", "response_ms": micro, "status_code": info["status"]
+    }
+
+
+def status(request):  # pylint: disable=unused-argument
+    """Status"""
+    info = {}
```
Comment (carsongee) [inline status/views.py:116]: So I was thinking, since connections are a limited resource, and these do "cost" us.  It would be nice to just do a simple query string type token that is defined in settings to protect this page (i.e. just return a 403 or 404).  Health checkers take a URL and all of the ones I've run into honor the `?=` querystring type params

File: .travis.yml (hunk 1 above)
Comment (ShawnMilo) [inline .travis.yml:10]: [Travis docs](http://docs.travis-ci.com/user/environment-variables/#Defining-Variables-in-.travis.yml)

Originally I did this:

``` yaml
env:
  - DATABASE_URL=postgres://postgres@localhost:5432/postgres
  - BROKER_URL=redis://localhost:6379/4
  - HAYSTACK_URL=127.0.0.1:9200
```

However, this doesn't work because Travis won't use all three at once. It does three runs using one each. So the first run through, there's no config for Redis or Elasticsearch.

This also works:

``` yaml
env:
  - DATABASE_URL=postgres://postgres@localhost:5432/postgres BROKER_URL=redis://localhost:6379/4 HAYSTACK_URL=127.0.0.1:9200
```

Both have the same effect, but I think the one-liner is uglier.

File: status/views.py (hunk 3 above)
Comment (ShawnMilo) [inline status/views.py]: I don't understand what is desired here. The values are being put into a shared dictionary, and I don't see where the potential is to reduce error. Please elaborate.

File: status/views.py (hunk 4 above)
Comment (ShawnMilo) [inline status/views.py:49]: I agree with the context manager, but would like argue on `timeit`. It moves code into strings to be `eval`ed, which feels dirty for production code. It looks like it's meant for benchmarks, not for production use. Increasing resolution from microseconds is unnecessary anyway, as network latency will be a far bigger factor in variation than anything else.

File: status/views.py (hunk 5 above)
Comment (ShawnMilo) [inline status/views.py:42]: Are you saying that if everything else is functional, but Redis is down, you want a 500 error, and that you'd rather not know that the problem was Redis (or PostgreSQL or Elasticsearch)?

It seems that "something is broken" is not useful, and using one of the many free monitoring services would serve that purpose nicely. There's no point in this branch at all if that's the case. Knowing that the Elasticsearch server is down would give a much bigger head-start on fixing the problem.

File: status/views.py (hunk 7 above)
Comment (ShawnMilo) [inline status/views.py:69]: Efforts were made here to directly connect to PostgreSQL, bypassing the ORM, to Redis, bypassing Celery, and to Elasticsearch, bypassing Haystack, for a true check on the status of the backend servers. This is better than using the intermediary, and is also useful for when a server (such as Redis) is being used by more than one plugin.

File: status/views.py (hunk 8 above)
Comment (ShawnMilo) [inline status/views.py]: I tried this and it doesn't work. If you have a working example I'll change it.

As for changing to solr, I think that's YAGNI until we do.

File: status/views.py (hunk 9 above)
Comment (ShawnMilo) [inline status/views.py:85]: Great point. I will close the connections.

File: status/views.py (hunk 11 above)
Comment (ShawnMilo) [inline status/views.py:116]: Okay, cool. The connection is going to be SSL everywhere, right? Just wanting to make sure the querystring can't be intercepted.

File: status/views.py (hunk 4 above)
Comment (ShawnMilo) [inline status/views.py:49]: After adding a context manager, it results in more complexity than it's worth. The context manager returns a value on `__enter__`, not `__exit__`, but it must have an `__exit__` method or it's invalid. The timer would be kicked off in the `__enter_`, and that's where the value must be returned from, so it gets ugly and it seems like this isn't actually a good use for a context manager.

If you think I'm missing the approach you had in mind, please send a link to a context manager that works the way you were thinking and I'll give it another go.

File: status/views.py (hunk 3 above)
Comment (carsongee) [inline status/views.py]: a `STATUSES` var at the top, with `no config found` (except that one isn't needed I don't think), `up`, and `down`. Then you can just use that everywhere you need to set status.

File: status/views.py (hunk 4 above)
Comment (carsongee) [inline status/views.py:49]: I was thinking you would use a dictionary with a `start` and `stop` that would register the times, and then that could be used outside the context, i.e.

``` python
@contextmanager
def timer(thing):
     try:
         thing['start'] = 1
         yield thing
     finally:
         thing['stop'] = 2
```

and then called like:

``` python
timer = {}
with timer(timer):
    ...
print(timer['stop'] - timer['start'])
```

but probably could be a lot smarter than that.

File: status/views.py (hunk 7 above)
Comment (carsongee) [inline status/views.py:69]: I agree on the bypassing, I just mean that they already probably have a function to parse the URL

File: status/views.py (hunk 8 above)
Comment (carsongee) [inline status/views.py]: Fair, on typing, you can skip that. You just need to pass it as:

``` python
Elasticsearch([settings.HAYSTACK_URL])
```

from http://elasticsearch-py.readthedocs.org/en/master/api.html#elasticsearch

Real world example: https://github.com/carsongee/archelon/blob/master/archelond/archelond/tests/config/elastic.py#L17 and https://github.com/carsongee/archelon/blob/master/archelond/archelond/data/elastic.py#L39-L41

File: status/views.py (hunk 4 above)
Comment (carsongee) [inline status/views.py:49]: That is cool on timeit, millis are probably fine, let alone micro.

**Comment by ShawnMilo** [comment]
Thanks for the updates, @carsongee. Is this ready to be moved back to "Waiting on Author"?

**Comment by carsongee** [comment]
Sorry, yeah. I forgot to flip the flag

File: status/views.py (hunk 5 above)
Comment (carsongee) [inline status/views.py:42]: I want both, so a 500 result returned, and the JSON with the details, that way the Load Balancer knows the service is unhealthy (say only one of 3 instances can't connect to one of these services, it should be removed from the LB), but our more sophisticated monitoring (i.e. me, or a script designed to use this page) will also want the full status JSON to know of a specific instance. So I guess we don't want to raise, but we do need to still send a 500

File: status/views.py (hunk 7 above)
Comment (carsongee) [inline status/views.py:69]: To be more specific: https://github.com/celery/celery/blob/a7c168d6df82bcd060e02e647a916d61627959dc/celery/backends/redis.py#L59 and use connparams from that instantiated class to do the connection or just use its client property.

File: .travis.yml (hunk 1 above)
Comment (carsongee) [inline .travis.yml:10]: Cool, thanks for the explanation.

File: status/views.py (hunk 5 above)
Comment (ShawnMilo) [inline status/views.py:42]: As discussed in meatspace, we're going to return a [503](http://www.checkupdown.com/status/E503.html) status code and the JSON.

Comment (ShawnMilo) [inline status/views.py:42]: This is now addressed in the code & tests.

File: status/views.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+    try:
+        url = settings.HAYSTACK_CONNECTIONS["default"]["URL"]
+    except (AttributeError, KeyError):
+        log.error("No Elasticsearch connection info in settings.")
+        return {"status": NO_CONFIG}
+    search = Elasticsearch(url)
+    start = datetime.now()
+    try:
+        info = search.info()
+    except ESConnectionError:
```
Comment (carsongee) [inline status/views.py:101]: With ES turned off in docker-compose, this just causes it to seemingly infinitely retry, so the page never renders (and would hit the load balancer timeout if it was in production)

File: status/views.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+        match = None
+    if match is None:
+        log.error("Invalid Redis connection info found in settings: %s", url)
+        return {"status": NO_CONFIG}
+    host, port, database = match.groups()
+    rdb = StrictRedis(host=host, port=port, db=database)
+    start = datetime.now()
+    try:
+        info = rdb.info()
+    except RedisConnectionError:
```
Comment (carsongee) [inline status/views.py]: I appear to get the same result with Redis, removing the redis link in docker-compose appears to cause the page to infinitely try.

Comment (carsongee) [inline status/views.py]: It came back after 2 minutes with the correct response, but that is 1.5 minutes more than the current LB timeout.

File: status/views.py (hunk 12 above)
Comment (ShawnMilo) [inline status/views.py:101]: I'm also getting a two-minute timeout with `psycopg2` during one test. However, there don't seem to be any timeout kwargs available for any of those libraries.

A quick Google search suggests the best solution is the [signals](https://docs.python.org/2/library/signal.html) package.

**Comment by pwilkins** [comment]
@ShawnMilo Are these test failures significant and you have more work to do? Should I flip the switch back to "Waiting on Author"?

**Comment by ShawnMilo** [comment]
Yeah, it looks like the failure on Python 3 is legit. I'll fix locally and re-push.

**Comment by ShawnMilo** [comment]
Tests are all passing locally now; not to see if Travis agrees.

File: status/views.py [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+    try:
+        url = settings.BROKER_URL
+        _, host, port, _, _, db, _ = parse_redis_url(url)
+    except AttributeError:
+        log.error("No valid Redis connection info found in settings.")
+        return {"status": NO_CONFIG}
+
+    start = datetime.now()
+    try:
+        with Timeout(2):
```
Comment (carsongee) [inline status/views.py]: Can we make this configurable with a default, or at least a constant at the top of the file to make it easier to change later if 2 seconds proves to be problematic?

**Comment by carsongee** [comment]
Got this with redis turned off:
image

**Comment by carsongee** [comment]
Running off of uwsgi instead of the development server appears to resolve the issue and works as expected.  If you can't figure out an easy way around that, I think this is still probably good enough after the magic number fix mentioned.

**Comment by ShawnMilo** [comment]
Magic number & merge conflict with `master` addressed. Still needs flattening if it passes review. I don't have any solution for the signal thing since _it_ is the only solution I found for our other problem.

Specifically, since the Redis, Elasticsearch, and PostgreSQL drivers don't have any timeout kwargs available, I don't know how else to handle it other than letting them time out. I'm open to suggestions.

**Comment by carsongee** [comment]
There is one for redis 2.10 and  greater: https://github.com/andymccurdy/redis-py/blob/81429b44acff3f238a212a1651241b49cde76c85/redis/connection.py#L388, and it looks like elasticsearch can do that as well: https://elasticsearch-py.readthedocs.org/en/master/api.html?highlight=timeout#timeout, and for psycopg2 as well: http://stackoverflow.com/questions/27641740/python-psycopg2-timeout

Though that is just from some googling, your mileage may vary.

**Comment by ShawnMilo** [comment]
I Googled those things a couple of weeks ago and came up short. My Google-fu must be wanting.

**Comment by ShawnMilo** [comment]
Timeout args added, `signals` removed. Tests currently passing.

**Comment by carsongee** [comment]
:purse: has an extra merge commit in there, but feel free to merge after commit is clean, and tests pass. Also, thanks.  This is really cool and I really learned a lot, sorry it took so long.

**Comment by ShawnMilo** [comment]
Excellent. Thanks for holding it to your usual high standards. I learned some cool stuff too. Like you can return data with 500 errors. :o)
